### ISSUE TITLE
Reconsider architecure of notebook loggers

### ISSUE THREAD

**Issue by greazer** [microsoft/vscode-jupyter#2192]
In [this PR](https://github.com/microsoft/vscode-python/pull/9755#), it was suggested that we reconfigure how notebook loggers and their related components are managed by the extension. 

Here is a copy of what was suggested.

Still not convinced this should be exposed.
We're exposing this just to get access to one member in the list. Feels wrong to expose everything.

Why not create a class `IGatherProviderRegistry` (ignore the name for now)
```typescript
IGatherProviderRegistry {
    private notebookLoggers = new Map<INotebook, INotebookExecutionLogger[]>();
    register(nb: INotebook, loggers: INotebookExecutionLogger[]){
		this.notebookLoggers.set(nb, loggers);
    }
    getProvider(nb: INotebook){
        // Go through loggers for this nb and return the object
		const loggers = this.notebookLoggers.get(nb);
        return loggers.find((logger: INotebookExecutionLogger) => {
      		// Find the logger that implements IGatherLogger or is a specific type..
     		return true // else false;
        });
    }
	// This is an overloaded method.
    getProvider(uri: Uri){
		// Given the Uri, we can get the notebook from the notebook provider class.
		// Now that we have a notebook we can get the provider.
    }
}
```

```typescript
// Register IGatherProviderRegistry as a singleton
```

```typescript
// In the contructor of jupyterNotebook (jupyterNotebook.ts)
constructor(gatherProviderRegistry: IGatherProviderRegistry){
	gatherProviderRegistry.register(this, this.loggers);
}
```

```typescript
// In other parts of the code you can
// Without any loggers, providers and the like.
gatherProviderRegistry.getProvider(notebook or notebookUri)
```

This way we won't have to modify INotebook, and the coupling is reduced, and all of this logic can sit elsewhere.

Basically we let IGatherProviderRegistry figure out the logic used to get the provider for a notebook.
We have code in a number of places that do this, from loggers to gather provider, then gather provider to `getService`...
We can move all of that into one place, then we don't need loggers, we don't need gather provider, we just need one class `IGatherProviderRegistry` which is kind of like a map that indexes using the nb or the uri.

Finally currently codeLensFactory does a bunch of setup for the cellHash infrastructure. That code really doesn't belong there and should be modeled (at least) more like the gather-related code.
